### PR TITLE
feat: Make snapshotID optional for volumeInitializationRate when AMI is selected via amiSelectorTerms

### LINKED ISSUE #8389: Make snapshotID optional for volumeInitializationRate when AMI is selected via amiSelectorTerms
### Description

**What problem are you trying to solve?**
I want to use volumeInitializationRate in Karpenter to make root EBS volume initialization predictable and faster when nodes launch from AMIs selected dynamically via amiSelectorTerms (tag-based selection). However, volumeInitializationRate currently appears to be allowed only when snapshotID is explicitly set in spec.blockDeviceMappings. This forces me to hardcode a specific snapshot, which breaks the whole point of dynamic AMI selection.

Concretely:
	•	We select AMIs by tags/alias using amiSelectorTerms so that Karpenter always picks the latest compatible AMI.  ￼
	•	We also want to set volumeInitializationRate on the root device to speed up initialization. But that requires snapshotID, tying the NodeClass to one snapshot and defeating dynamic AMI rotation.

**How important is this feature to you?**
High. We regularly roll AMIs forward. Manually updating snapshotID each time creates drift, increases operational toil, and risks mismatches between AMI and snapshot during deployments. Enabling volumeInitializationRate without losing dynamic AMI selection would improve boot predictability while keeping our AMI management safe and hands-off.

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by atilsensalduz**
Fixes #8389 

**Description**

This PR implements the enhancement requested in #8389 to make `snapshotID` optional for `volumeInitializationRate` when AMI is selected via `amiSelectorTerms`. This enables users to benefit from faster EBS volume initialization while using dynamic AMI selection, reducing operational overhead.

Currently, Karpenter requires an explicit `snapshotID` when using `volumeInitializationRate` in block device mappings. This prevents dynamic AMI selection via `amiSelectorTerms` while still benefiting from faster EBS volume initialization, creating operational overhead for users who want to manage AMIs dynamically.

## Solution

Updated the CEL validation rule to allow `volumeInitializationRate` without explicit `snapshotID` when the EC2NodeClass has `amiSelectorTerms` configured. When AMIs are selected dynamically, they contain block device mapping information with snapshots that AWS can use for volume initialization.

**How was this change tested?**
```
go test ./pkg/apis/v1 -run "TestCEL.*BlockDeviceMappings" -v
```

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates 
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jmdeal** [comment]
Do we need this validation at all? Users always need to specify via `amiSelectorTerms`, there's no other way to select an AMI in Karpenter. The one edge-case where I could see this not being supported is when you're using an AMI which uses the instance's local storage rather than EBS as the root volume ([ref](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/creating-an-ami-instance-store.html)). There's no way for us to detect the type of AMI selected at validation time, so we can't perform a CEL validation against that.

Given so few instances that support that edge-case anyway and you would have to explicitly opt-in to encounter an issue (by configuring volumeInitializationRate) it seems reasonable to not perform this validation at all. I'd need to test locally to see if there are any other edge cases I'm unaware of - I'm not an expert on this EBS feature.

**Comment by jmdeal** [comment]
Actually, there is one other thing we need to handle here - even though the snapshot ID for the root device is derived implicitly from the AMI, you still can't create a launch template that only specifies `VolumeInitializationRate` and use it with `ec2:RunInstances` - you get and `InvalidParameterCombination` error. I imagine all we should need to do is copy the snapshot ID for the AMI's root block device (discovered via DescribeImages) and pipe that into our generated launch template. That shouldn't be too difficult but does expand the scope of this PR.

**Comment by moko-poi** [comment]
Hey @atilsensalduz,

Just checking in on this PR - any thoughts on @jmdeal's feedback from December? Particularly regarding the `InvalidParameterCombination` error and potentially piping the snapshot ID from the AMI's root block device?

Let me know if you're still working on this or need any assistance!

**Comment by atilsensalduz** [comment]
Hi @jmdeal, apologies for missing your earlier comments.

You're right. The current Karpenter version mandates amiSelectorTerms, so the CEL validation would always pass. So, I've removed it.

I've also adjusted the PR based on your suggestion. Right now, if a user specifies volumeInitializationRate on their root volume without providing an explicit snapshotID, we now inject the AMI's snapshot ID automatically. This way users don't need to manually look up and specify the snapshot ID anymore. Karpenter handles it transparently.

Thank you so much for the heads-up about the InvalidParameterCombination error. That was really helpful!

Let me know if you have any concerns or if I've missed anything else.

**Comment by atilsensalduz** [comment]
Hi @moko-poi , thank you for catching this. I've addressed @jmdeal's feedbacks in the latest changes. Could you please take a look at the recent changes when you get a chance?

**Comment by atilsensalduz** [comment]
Tested locally by running the controller against a real EKS cluster (make run) with the following EC2NodeClass:

```
apiVersion: karpenter.k8s.aws/v1
kind: EC2NodeClass
metadata:
  name: test-vol-init
spec:
  amiSelectorTerms:
    - alias: al2023@latest
  subnetSelectorTerms:
    - tags:
        karpenter.sh/discovery: "<cluster-name>"
  securityGroupSelectorTerms:
    - tags:
        karpenter.sh/discovery: "<cluster-name>"
  role: karpenter-role-<cluster-name>
  blockDeviceMappings:
    - deviceName: /dev/xvda
      rootVolume: true
      ebs:
        volumeSize: 50Gi
        volumeInitializationRate: 200   # no snapshotID
```

The controller successfully resolved all AMIs and automatically populated rootDeviceSnapshotID in the status for each, confirming the snapshot ID injection works as expected:


```
apiVersion: karpenter.k8s.aws/v1
kind: EC2NodeClass
metadata:
  annotations:
    karpenter.k8s.aws/ec2nodeclass-hash: "<hash>"
    karpenter.k8s.aws/ec2nodeclass-hash-version: v4
    kubectl.kubernetes.io/last-applied-configuration: "<redacted>"
  creationTimestamp: "<timestamp>"
  finalizers:
  - karpenter.k8s.aws/termination
  generation: 1
  name: test-vol-init
  resourceVersion: "<resourceVersion>"
  uid: "<uid>"
spec:
  amiSelectorTerms:
  - alias: al2023@latest
  blockDeviceMappings:
  - deviceName: /dev/xvda
    ebs:
      volumeInitializationRate: 200
      volumeSize: 50Gi
    rootVolume: true
  metadataOptions:
    httpEndpoint: enabled
    httpProtocolIPv6: disabled
    httpPutResponseHopLimit: 1
    httpTokens: required
  role: karpenter-role-<cluster-name>
  securityGroupSelectorTerms:
  - tags:
      karpenter.sh/discovery: <cluster-name>
  subnetSelectorTerms:
  - tags:
      karpenter.sh/discovery: <cluster-name>
status:
  amis:
  - id: ami-xxxxxxxxxxxxxxxxx
    name: amazon-eks-node-al2023-arm64-nvidia-1.32-v20260209
    requirements:
    - key: kubernetes.io/arch
      operator: In
      values:
      - arm64
    - key: karpenter.k8s.aws/instance-gpu-count
      operator: Exists
    rootDeviceName: /dev/xvda
    rootDeviceSnapshotID: snap-xxxxxxxxxxxxxxxxx
  - id: ami-xxxxxxxxxxxxxxxxx
    name: amazon-eks-node-al2023-x86_64-nvidia-1.32-v20260209
    requirements:
    - key: kubernetes.io/arch
      operator: In
      values:
      - amd64
    - key: karpenter.k8s.aws/instance-gpu-count
      operator: Exists
    rootDeviceName: /dev/xvda
    rootDeviceSnapshotID: snap-xxxxxxxxxxxxxxxxx
  - id: ami-xxxxxxxxxxxxxxxxx
    name: amazon-eks-node-al2023-x86_64-neuron-1.32-v20260209
    requirements:
    - key: kubernetes.io/arch
      operator: In
      values:
      - amd64
    - key: karpenter.k8s.aws/instance-accelerator-count
      operator: Exists
    rootDeviceName: /dev/xvda
    rootDeviceSnapshotID: snap-xxxxxxxxxxxxxxxxx
  - id: ami-xxxxxxxxxxxxxxxxx
    name: amazon-eks-node-al2023-arm64-standard-1.32-v20260209
    requirements:
    - key: kubernetes.io/arch
      operator: In
      values:
      - arm64
    - key: karpenter.k8s.aws/instance-gpu-count
      operator: DoesNotExist
    - key: karpenter.k8s.aws/instance-accelerator-count
      operator: DoesNotExist
    rootDeviceName: /dev/xvda
    rootDeviceSnapshotID: snap-xxxxxxxxxxxxxxxxx
  - id: ami-xxxxxxxxxxxxxxxxx
    name: amazon-eks-node-al2023-x86_64-standard-1.32-v20260209
    requirements:
    - key: kubernetes.io/arch
      operator: In
      values:
      - amd64
    - key: karpenter.k8s.aws/instance-gpu-count
      operator: DoesNotExist
    - key: karpenter.k8s.aws/instance-accelerator-count
      operator: DoesNotExist
    rootDeviceName: /dev/xvda
    rootDeviceSnapshotID: snap-xxxxxxxxxxxxxxxxx
  conditions:
  - lastTransitionTime: "<timestamp>"
    message: ""
    observedGeneration: 1
    reason: AMIsReady
    status: "True"
    type: AMIsReady
  - lastTransitionTime: "<timestamp>"
    message: ""
    observedGeneration: 1
    reason: CapacityReservationsReady
    status: "True"
    type: CapacityReservationsReady
  - lastTransitionTime: "<timestamp>"
    message: SubnetSelector did not match any Subnets
    observedGeneration: 1
    reason: SubnetsNotFound
    status: "False"
    type: SubnetsReady
  - lastTransitionTime: "<timestamp>"
    message: ""
    observedGeneration: 1
    reason: SecurityGroupsReady
    status: "True"
    type: SecurityGroupsReady
  - lastTransitionTime: "<timestamp>"
    message: ""
    observedGeneration: 1
    reason: InstanceProfileReady
    status: "True"
    type: InstanceProfileReady
  - lastTransitionTime: "<timestamp>"
    message: Awaiting AMI, Instance Profile, Security Group, and Subnet resolution
    observedGeneration: 1
    reason: DependenciesNotReady
    status: "False"
    type: ValidationSucceeded
  - lastTransitionTime: "<timestamp>"
    message: ValidationSucceeded=False, SubnetsReady=False
    observedGeneration: 1
    reason: UnhealthyDependents
    status: "False"
    type: Ready
  instanceProfile: <cluster-name>_<hash>
  securityGroups:
  - id: sg-xxxxxxxxxxxxxxxxx
    name: <cluster-name>-node-group-<hash>
```

File: pkg/apis/v1/ec2nodeclass_status.go [hunk 1]
```diff
@@ -77,6 +77,12 @@ type AMI struct {
 	// Requirements of the AMI to be utilized on an instance type
 	// +required
 	Requirements []corev1.NodeSelectorRequirement `json:"requirements"`
+	// RootDeviceName is the device name of the root volume (/dev/xvda)
```
Comment (Sarthug99) [inline pkg/apis/v1/ec2nodeclass_status.go]: nit : add eg. here matching the CRD description. /dev/xvda isn't the only possible root device name.

**Comment by Sarthug99** [comment]
One more thing to consider here - 
The old CEL rule rejected `volumeInitializationRate` without snapshot ID on **any** block device mapping. The new injection logic only covers root volumes (matched by device name or rootVolume flag). 
This means a non-root volume with `volumeInitializationRate` and no snapshot ID will now pass the validation and fail at the EC2 API. We should either keep the CEL validation for non-root volumes, or add validation in `injectSnapshotIDForVolumeInitialization` to address this scenario.
